lib/core/test/unit/Cardano/Wallet/DBSpec2.hs

```diff
@@ -0,0 +1,595 @@
+
+widPK :: MWid -> PrimaryKey WalletId
+widPK (MWid wid) = PrimaryKey wid
+
+pkWid :: PrimaryKey WalletId -> MWid
+pkWid (PrimaryKey wid) = MWid wid
+
+data Mock = M
+    { checkpoints :: Map MWid MWallet
+    , metas :: Map MWid WalletMetadata
+    , txs :: Map MWid TxHistory
+    , privateKey :: Map MWid MPrivKey
+    , next :: MConn
+    }
+    deriving (Show, Generic)
+
+emptyMock :: Mock
+emptyMock = M Map.empty Map.empty Map.empty Map.empty (MConn 0)
+
+type MockOp a = Mock -> (Either Err a, Mock)
+
+mOpen :: MockOp MConn
+mOpen (M cp metas txs pk n) = (Right n, M cp metas txs pk (succ n))
+
+mClose :: MConn -> MockOp ()
+mClose _ m = (Right (), m)
+
+mCreateWallet :: MWid -> MWallet -> WalletMetadata -> MockOp ()
+mCreateWallet wid wal meta m@(M cp metas txs pk n)
+    | wid `Map.member` cp = (Left (WalletAlreadyExists wid), m)
+    | otherwise           = (Right (), M (Map.insert wid wal cp) (Map.insert wid meta metas) txs pk n)
+
+mRemoveWallet :: MWid -> MockOp ()
+mRemoveWallet wid m@(M cp metas txs pk n)
+    | wid `Map.member` cp = (Right (), M (Map.delete wid cp) (Map.delete wid metas) txs pk n)
+    | otherwise           = (Left (NoSuchWallet wid), m)
+
+mListWallets :: MockOp [MWid]
+mListWallets m@(M cp _ _ _ _) = (Right (Map.keys cp), m)
+
+mPutCheckpoint :: MWid -> MWallet -> MockOp ()
+mPutCheckpoint wid wal m@(M cp metas txs pk n)
+    | wid `Map.member` cp = (Right (), M (Map.insert wid wal cp) metas txs pk n)
+    | otherwise           = (Left (NoSuchWallet wid), m)
+
+mReadCheckpoint :: MWid -> MockOp (Maybe MWallet)
+mReadCheckpoint wid m@(M cp _ _ _ _)
+    = (Right (Map.lookup wid cp), m)
+
+mPutWalletMeta :: MWid -> WalletMetadata -> MockOp ()
+mPutWalletMeta wid meta m@(M cp metas txs pk n)
+    | wid `Map.member` cp = (Right (), M cp (Map.insert wid meta metas) txs pk n)
+    | otherwise           = (Left (NoSuchWallet wid), m)
+
+mReadWalletMeta :: MWid -> MockOp (Maybe WalletMetadata)
+mReadWalletMeta wid m@(M _ meta _ _ _)
+    = (Right (Map.lookup wid meta), m)
+
+mPutTxHistory :: MWid -> TxHistory -> MockOp ()
+mPutTxHistory wid txs' m@(M cp metas txs pk n)
+    | wid `Map.member` cp = (Right (), M cp metas (Map.alter appendTxs wid txs) pk n)
+    | otherwise           = (Left (NoSuchWallet wid), m)
+    where appendTxs = Just . (<> txs') . fromMaybe mempty
+
+mReadTxHistory :: MWid -> MockOp TxHistory
+mReadTxHistory wid m@(M cp _ txs _ _)
+    | wid `Map.member` cp = (Right (fromMaybe mempty (Map.lookup wid txs)), m)
+    | otherwise           = (Right mempty, m)
+
+mPutPrivateKey :: MWid -> MPrivKey -> MockOp ()
+mPutPrivateKey wid pk' m@(M cp metas txs pk n)
+    | wid `Map.member` cp = (Right (), M cp metas txs (Map.insert wid pk' pk) n)
+    | otherwise           = (Left (NoSuchWallet wid), m)
+
+mReadPrivateKey :: MWid -> MockOp (Maybe MPrivKey)
+mReadPrivateKey wid m@(M cp _ _ pk _)
+    | wid `Map.member` cp = (Right (Map.lookup wid pk), m)
+    | otherwise           = (Left (NoSuchWallet wid), m)
```

Comment:
Wouldn't it be possible to leverage our already existing `MVar` implementation as the Mock, instead of re-encoding everything here :thinking: ? We've already some good unit tests which make sure that the MVar implementation works rather okay.